KAFKA-19212: Correct the unclean leader election metric calculation for 3.9

mmatloka opened the pull request:
This is attempt of backport https://github.com/apache/kafka/pull/19590 specific for Kafka 3.9. Copying description from that PR:

"""
The current ElectionWasClean checks if the new leader is in the previous
ISR. However, there is a corner case in the partition reassignment. The
partition reassignment can change the partition replicas. If the new
preferred leader (the first one in the new replicas) is the last one to
join ISR, this preferred leader will be elected in the same partition
change.

For example: In the previous state, the partition is Leader: 0,
Replicas (2,1,0), ISR (1,0), Adding(2), removing(0). Then replica 2
joins the ISR. The new partition would be like: Leader: 2, Replicas
(2,1), ISR(1,2). The new leader 2 is not in the previous ISR (1,0) but
it is still a clean election.
"""

The original fix depends on ELR, however, AFAIK ELR's are valid for Kafka 4.0 and 4.1, not for 3.9. I prepared a simpler fix for 3.9 (we have encountered such problem under 3.9, and Confluent confirmed it is an issue).

Note: this is my first Apache contribution, I signed the ICLA, I read the contributing guidelines, sorry if I did something incorrectly :) 
Let me also mention author of the original fix for 4.x - @CalvinConfluent 

Thank you!

chia7712 commented:
@mmatloka thanks for this patch. Do you have time to add an integration test to reproduce this bug?

mmatloka commented:
> @mmatloka thanks for this patch. Do you have time to add an integration test to reproduce this bug?

Hi, I will try to add additional tests

mmatloka commented:
> @mmatloka thanks for this patch. Do you have time to add an integration test to reproduce this bug?

Hi, I added some additional tests. Could you check them, and if they’re not sufficient, point out where I should add more (I am not fully familiar with this codebase)?

FrankYang0529 commented on metadata/src/main/java/org/apache/kafka/metadata/PartitionRegistration.java:
```diff
@@ -346,8 +346,8 @@ public String diff(PartitionRegistration prev) {
-    public void maybeLogPartitionChange(Logger log, String description, PartitionRegistration prev) {
-        if (!electionWasClean(leader, prev.isr)) {
+    public void maybeLogPartitionChange(Logger log, String description, PartitionRegistration prev, PartitionRegistration next) {
+        if (!electionWasClean(leader, prev.isr, next.addingReplicas)) {
```
Could we use `addingReplicas` directly? It's like `diff` function only needs `prev` argument.

mmatloka commented on metadata/src/main/java/org/apache/kafka/controller/metrics/ControllerMetricsChanges.java:
```diff
@@ -138,7 +142,9 @@ void handlePartitionChange(PartitionRegistration prev, PartitionRegistration nex
             int[] prevIsr = prev != null ? prev.isr : next.replicas;
-            if (!PartitionRegistration.electionWasClean(next.leader, prevIsr)) {
+            // check if at the same step the partition we are adding to ISR is becoming the leader, don't treat that as unclean election.
+            int[] nextAddingReplicas = next.addingReplicas;
```
Could someone doublecheck my thinking? AddingReplicas will appear in prev or in next?

gaurav-narula commented:
Thanks for the patch! I've a question:

> The original fix depends on ELR

I'm unsure about this. Going through the original fix, IIUC, it modifies the "unclean" check to be based on `LeaderRecoveryState.RECOVERING` [[0](https://github.com/apache/kafka/blob/95d164a3ae67b8d3860567528e2ba6ca0c532fa2/metadata/src/main/java/org/apache/kafka/metadata/PartitionRegistration.java#L166)] which was introduced with [KIP-704](https://cwiki.apache.org/confluence/display/KAFKA/KIP-704%3A+Send+a+hint+to+the+partition+leader+to+recover+the+partition) and as a result should be available with 3.9. It also modifies the calculation of `electionFromElrCounter` which was introduced in 4.1 with [KAFKA-18954](https://issues.apache.org/jira/browse/KAFKA-18954) but that I suspect that can be dropped from the backport?

mmatloka commented:
> Thanks for the patch! I've a question:
> 
> > The original fix depends on ELR
> 
> I'm unsure about this. Going through the original fix, IIUC, it modifies the "unclean" check to be based on `LeaderRecoveryState.RECOVERING` [[0](https://github.com/apache/kafka/blob/95d164a3ae67b8d3860567528e2ba6ca0c532fa2/metadata/src/main/java/org/apache/kafka/metadata/PartitionRegistration.java#L166)] which was introduced with [KIP-704](https://cwiki.apache.org/confluence/display/KAFKA/KIP-704%3A+Send+a+hint+to+the+partition+leader+to+recover+the+partition) and as a result should be available with 3.9. It also modifies the calculation of `electionFromElrCounter` which was introduced in 4.1 with [KAFKA-18954](https://issues.apache.org/jira/browse/KAFKA-18954) but that I suspect that can be dropped from the backport?

Hi, actually the good question is what events actually lead to the situation where adding replica becomes the leader in one step. One situation is probably when it is recovering. However, what we saw in practice, that the situation was probably(?) result of some partition re-assign (probably performed by Confluent SBC actually, new cluster, completely healthy, idle, nothing happening, suddenly adding partition becomes the leader and metrics show unclean election. maybe manual reassign could try to simulate this 🤔 ). Do the partition re-assign cause the leader go through recovering state?

FrankYang0529 commented:
@mmatloka I agree with @gaurav-narula. Based on https://github.com/apache/kafka/pull/19590, it uses `LeaderRecoveryState.RECOVERING` to correct the metric. The `LeaderRecoveryState.RECOVERING` is used in two cases.
1. The election result is unclean. If the leader is in `targetIsr`, the result cannot be unclean.
https://github.com/apache/kafka/blob/506ff18e8887cf9610489f56f643a081d015394e/metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java#L357-L366
2. Elect a leader for offline. This will trigger `leaderAndIsr.newRecoveringLeaderAndIsr` and get `LeaderRecoveryState.RECOVERING`. However, for a new partition you mentioned, its state is changed from `NewPartition` to `OnlinePartition`, so it doesn't get into this path.
https://github.com/apache/kafka/blob/506ff18e8887cf9610489f56f643a081d015394e/core/src/main/scala/kafka/controller/Election.scala#L28-L58

Could you help to cherry-pick https://github.com/apache/kafka/commit/3094ce2c202c96c57084e82892df44b1f124c481 and ignore `electionFromElr`? Thanks.

mmatloka committed:
Backport KAFKA-19212: Correct the unclean leader election metric calculation (#19590)

The current ElectionWasClean checks if the new leader is in the previous
ISR. However, there is a corner case in the partition reassignment.  The
partition reassignment can change the partition replicas. If the new
preferred leader (the first one in the new replicas) is the last one to
join ISR, this preferred leader will be elected in the same partition
change.

For example:  In the previous state, the partition is  Leader: 0,
Replicas (2,1,0), ISR (1,0), Adding(2), removing(0).  Then replica 2
joins the ISR. The new partition would be like:  Leader: 2, Replicas
(2,1), ISR(1,2).  The new leader 2 is not in the previous ISR (1,0) but
it is still a clean election.

Reviewers: Jun Rao <junrao@gmail.com>
(cherry picked from commit 3094ce2c202c96c57084e82892df44b1f124c481)

# Conflicts:
#	metadata/src/main/java/org/apache/kafka/controller/metrics/ControllerMetricsChanges.java
#	metadata/src/main/java/org/apache/kafka/metadata/PartitionRegistration.java
#	metadata/src/test/java/org/apache/kafka/metadata/PartitionRegistrationTest.java

mmatloka commented:
> @mmatloka I agree with @gaurav-narula. Based on #19590, it uses `LeaderRecoveryState.RECOVERING` to correct the metric. The `LeaderRecoveryState.RECOVERING` is used in two cases.
> 
> 1. The election result is unclean. If the leader is in `targetIsr`, the result cannot be unclean.
>    https://github.com/apache/kafka/blob/506ff18e8887cf9610489f56f643a081d015394e/metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java#L357-L366
> 2. Elect a leader for offline. This will trigger `leaderAndIsr.newRecoveringLeaderAndIsr` and get `LeaderRecoveryState.RECOVERING`. However, for a new partition you mentioned, its state is changed from `NewPartition` to `OnlinePartition`, so it doesn't get into this path.
>    https://github.com/apache/kafka/blob/506ff18e8887cf9610489f56f643a081d015394e/core/src/main/scala/kafka/controller/Election.scala#L28-L58
> 
> Could you help to cherry-pick [3094ce2](https://github.com/apache/kafka/commit/3094ce2c202c96c57084e82892df44b1f124c481) and ignore `electionFromElr`? Thanks.

Done, please take a look. Thank you!

FrankYang0529 reviewed (approved):
LGTM. Thanks for the patch. Failed cases cannot be reproduced on my laptop.

mmatloka commented:
How to get second approval? :) @chia7712 @gaurav-narula ?

mmatloka commented:
> How to get second approval? :) @chia7712 @gaurav-narula ?

hey @FrankYang0529  can yo help here? :)

FrankYang0529 commented:
@chia7712 could you have the final check on this? Thank you.

chia7712 commented:
Based on my reading, this issue only causes incorrect metrics and does not impact core functionality. I'm hesitant to include this in 3.9, particularly if it is no even in 4.0

mmatloka commented:
> Based on my reading, this issue only causes incorrect metrics and does not impact core functionality. I'm hesitant to include this in 3.9, particularly if it is no even in 4.0

Hi @chia7712 . It is a situation where metrics spawn incorrectly alerts related to unclean leader election. How to determine if the unclean election is happening with this bug in 3.9?

chia7712 commented:
> Hi @chia7712 . It is a situation where metrics spawn incorrectly alerts related to unclean leader election. How to determine if the unclean election is happening with this bug in 3.9?

The incorrect metrics are triggered by partition reassignment. So, it is likely a true alert if you didn't trigger any reassignment

mmatloka commented:
> > Hi @chia7712 . It is a situation where metrics spawn incorrectly alerts related to unclean leader election. How to determine if the unclean election is happening with this bug in 3.9?
> 
> The incorrect metrics are triggered by partition reassignment. So, it is likely a true alert if you didn't trigger any reassignment

Hi, could you confirm one thing. Can unclean leader election happen at all, in any possible edge case, when it is explicitly disabled in the settings?

About reassignment - this gets complicated on Confluent with self balancing which can spawn reassignments.
